### PR TITLE
Allow static linking with Z3

### PR DESCRIPTION
**Pull request by ekpyron**
Experimental attempt to build an entirely static version of ``solc`` on circleci.

### CODE REVIEW

**Comment by ekpyron** [comment]
Wow - the static binary with debug information is 380MB now... but well, stripped it's only 23MB and it *does* run on trusty and includes Z3. It should basically run everywhere.
However, the build is still quite hacky (updated cmake, custom built Z3, etc.).

File: libsolidity/CMakeLists.txt [hunk 1]
```diff
@@ -29,7 +29,15 @@ add_library(solidity ${sources} ${headers})
 target_link_libraries(solidity PUBLIC evmasm devcore ${Boost_FILESYSTEM_LIBRARY} ${Boost_SYSTEM_LIBRARY})
 
 if (${Z3_FOUND})
-  target_link_libraries(solidity PUBLIC ${Z3_LIBRARY})
+  if(SOLC_LINK_STATIC)
+	find_package(Threads)
+    set(Z3_LIBRARIES_EXTRA ${Z3_LIBRARIES_EXTRA} Threads::Threads)
+    find_package(OpenMP)
+	if(OpenMP_FOUND)
+		set(Z3_LIBRARIES_EXTRA ${Z3_LIBRARIES_EXTRA} OpenMP::OpenMP_CXX)
```
Comment (ekpyron) [inline libsolidity/CMakeLists.txt]: Unfortunately, this only works starting from cmake 3.9 - with older cmake it'd be even messier...

File: .circleci/config.yml [hunk 2]
```diff
@@ -139,6 +139,37 @@ jobs:
...
+      - run: mkdir -p /root/.local/bin
+      - run: echo 'export PATH=/root/.local/bin:$PATH' >> $BASH_ENV
+      - run:
+          name: Install build dependencies
+          command: |
+            apt-get -qq update
+            apt-get -qy install libboost-regex-dev libboost-filesystem-dev libboost-test-dev libboost-system-dev libboost-program-options-dev
+            ./scripts/install_cmake.sh
+            ./scripts/install_obsolete_jsoncpp_1_7_4.sh
+            ./scripts/install_static_z3.sh
```
Comment (ekpyron) [inline .circleci/config.yml]: By default Z3 doesn't ship static libs, so we have to build them ourselves...

**Comment by chriseth** [comment]
Oh by the way, we have the static build that runs on travis...

**Comment by ekpyron** [comment]
We do? Ah! In ``scripts/Dockerfile`` via ``scripts/docker_build.sh`` via ``.travis.yml`` - hadn't seen that :-). ~~That might still dynamically linked against ``libgcc`` and ``libstdc++``, though, and~~ it'll most probably not have Z3 support  - I'll have a look at it.

**Comment by chriseth** [comment]
What is the status here?

**Comment by axic** [comment]
Needs a good old rebase. It seemed the tests were passing.

**Comment by ekpyron** [comment]
I just rebased and squashed. But yes, it looks like this works.

**Comment by axic** [comment]
Can you also build cvc4? :)

**Comment by ekpyron** [comment]
@axic Still working on that - so far I ended up with segfaults when statically linking against cvc4 locally. In general it looks like building a static version of cvc4 is more messy than for z3 unfortunately.

**Comment by axic** [comment]
Haha.

Can you change the title though?

**Comment by ekpyron** [comment]
I'm not sure whether we want to have this as a CI run on every PR or rather just as a nightly one. Originally the idea of this PR was that we could drop trusty builds, since the executable built in this PR will run on trusty.

**Comment by axic** [comment]
You could also cache the static z3 build on circleci. @chfast do you have any pointers?

**Comment by chfast** [comment]
> You could also cache the static z3 build on circleci. @chfast do you have any pointers?

https://circleci.com/docs/2.0/caching/

**Comment by ekpyron** [comment]
I'll have a look!

**Comment by axic** [comment]
Alternatively we could move on to use the aleth build docker image and include z3 in that.

File: libsolidity/CMakeLists.txt [hunk 3]
```diff
@@ -29,7 +29,16 @@ add_library(solidity ${sources} ${headers})
 target_link_libraries(solidity PUBLIC evmasm devcore ${Boost_FILESYSTEM_LIBRARY} ${Boost_SYSTEM_LIBRARY})
 
 if (${Z3_FOUND})
-  target_link_libraries(solidity PUBLIC ${Z3_LIBRARY})
+  if(SOLC_LINK_STATIC)
```
Comment (axic) [inline libsolidity/CMakeLists.txt:80]: I think this changeset could be merge in an individual PR with a changelog entry ("z3 can now be statically linked"). This doesn't depend on this PR having circleci caching.

**Comment by ekpyron** [comment]
I re-purposed this to only allow static linking with Z3 without a CI run.

~~I still kept the build script, though~~ Moving forward I'd like to create a static PPA build of Z3 in our new static PPA repo instead - that could be used both for our test runs and as a build dependency for our static PPA builds. ~~For that I'd modify the install_static_z3 script to a release_static_z3_ppa script or something like that.~~ (EDIT: I removed the script after all)

File: scripts/install_static_z3.sh [hunk 4]
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env bash
+
+git clone --depth 1 --branch z3-4.8.1 https://github.com/Z3Prover/z3.git
+cd z3
+mkdir build
+cd build
+LDFLAGS="-static" cmake -DBUILD_LIBZ3_SHARED=OFF ..
```
Comment (ekpyron) [inline scripts/install_static_z3.sh]: If we don't want to keep the script, I could instead just make a note to myself that this is how to do it...

Comment (ekpyron) [reply scripts/install_static_z3.sh]: Removed it and created an issue with a note instead.

**Comment by chriseth** [comment]
Closing for now to be re-done properly.

**Pull request closed without merging**